sql: Aggregate functions should not be allowed in WHERE clauses

Currently a select statement like `SELECT x from t WHERE AVG(x) > 1;` will crash with a nil pointer deference. These statements should not be allowed, and should produce an error like `"aggregate functions are not allowed in WHERE"`. This may be related to #4140, but I foresee the correct fix being different.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4786b01]

goroutine 111 [running]:
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Eval(0xc8205c6300, 0x1, 0xece45a461, 0x2a056e2, 0x5971560, 0xece45a461, 0x2a056e2, 0x5971560, 0xc82029eec0, 0xc82053d5a0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1142 +0x541
github.com/cockroachdb/cockroach/sql/parser.(*ComparisonExpr).Eval(0xc8205aa4e0, 0x1, 0xece45a461, 0x2a056e2, 0x5971560, 0xece45a461, 0x2a056e2, 0x5971560, 0xc82029eec0, 0xc82053d5a0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1062 +0x8c
github.com/cockroachdb/cockroach/sql.(*selectNode).filterRow(0xc82007e900, 0xc82053d9e0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/select.go:464 +0xae
github.com/cockroachdb/cockroach/sql.(*selectNode).Next(0xc82007e900, 0xc82010f590)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/select.go:119 +0x1b6
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func1(0xece45a461, 0x2a056e2, 0x5971560, 0xc8205aa501, 0x28)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:452 +0x36a
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt.func2(0xc82015e000, 0x1)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:497 +0x121
```

--- a/sql/group.go
+++ b/sql/group.go
@@ -44,6 +44,11 @@ var aggregates = map[string]func() aggregateImpl{
 // groupBy constructs a groupNode according to grouping functions or clauses. This may adjust the
 // render targets in the selectNode as necessary.
 func (p *planner) groupBy(n *parser.Select, s *selectNode) (*groupNode, *roachpb.Error) {
+	// Make sure there are no aggregation functions in the select's WHERE clause.
+	if p.aggregateInWhere(n.Where) {
+		return nil, roachpb.NewUErrorf("aggregate functions are not allowed in WHERE")
+	}
+
 	// Determine if aggregation is being performed. This check is done on the raw
 	// Select expressions as simplification might have removed aggregation
 	// functions (e.g. `SELECT MIN(1)` -> `SELECT 1`).
@@ -506,11 +511,27 @@ func (v *isAggregateVisitor) Visit(expr parser.Expr, pre bool) (parser.Visitor, 
 	return v, expr
 }
 
+func (v *isAggregateVisitor) reset() {
+	v.aggregated = false
+}
+
+func (p *planner) aggregateInWhere(where *parser.Where) bool {
+	if where != nil {
+		defer p.isAggregateVisitor.reset()
+		_ = parser.WalkExpr(&p.isAggregateVisitor, where.Expr)
+		if p.isAggregateVisitor.aggregated {
+			return true
+		}
+	}
+	return false
+}
+
 func (p *planner) isAggregate(n *parser.Select) bool {
 	if n.Having != nil || len(n.GroupBy) > 0 {
 		return true
 	}
 
+	defer p.isAggregateVisitor.reset()
 	for _, target := range n.Exprs {
 		_ = parser.WalkExpr(&p.isAggregateVisitor, target.Expr)
 		if p.isAggregateVisitor.aggregated {

--- a/sql/testdata/aggregate
+++ b/sql/testdata/aggregate
@@ -181,6 +181,9 @@ SELECT COUNT(*), k+v FROM kv GROUP BY v
 query error column "v" must appear in the GROUP BY clause or be used in an aggregate function
 SELECT COUNT(*), v/(k+v) FROM kv GROUP BY k+v
 
+query error aggregate functions are not allowed in WHERE
+SELECT k FROM kv WHERE AVG(k) > 1
+
 # Test case from #2761.
 query II rowsort
 SELECT count(kv.k) AS count_1, kv.v + kv.w AS lx FROM kv GROUP BY kv.v + kv.w

